cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -52,31 +52,37 @@ SparsePoseGraph::~SparsePoseGraph() {
 }
 
 void SparsePoseGraph::GrowSubmapTransformsAsNeeded(
-    const std::vector<const mapping::Submap*>& submaps) {
-  CHECK(!submaps.empty());
-  CHECK_LT(submap_transforms_.size(), std::numeric_limits<int>::max());
-  const int next_transform_index = submap_transforms_.size();
+    const std::vector<const mapping::Submap*>& insertion_submaps) {
+  CHECK(!insertion_submaps.empty());
+  CHECK_LT(optimization_problem_.submap_data().size(),
+           std::numeric_limits<int>::max());
+  const int next_submap_index = optimization_problem_.submap_data().size();
   // Verify that we have an index for the first submap.
-  const int first_transform_index = GetSubmapIndex(submaps[0]);
-  if (submaps.size() == 1) {
+  const int first_submap_index = GetSubmapIndex(insertion_submaps[0]);
+  CHECK_LE(first_submap_index, next_submap_index);
+  if (insertion_submaps.size() == 1) {
     // If we don't already have an entry for this submap, add one.
-    if (first_transform_index == next_transform_index) {
-      submap_transforms_.push_back(transform::Rigid2d::Identity());
+    if (first_submap_index == next_submap_index) {
+      optimization_problem_.AddSubmap(
+          submap_states_[first_submap_index].trajectory,
+          transform::Rigid2d::Identity());
     }
     return;
   }
-  CHECK_EQ(2, submaps.size());
+  CHECK_EQ(2, insertion_submaps.size());
   // CHECK that we have a index for the second submap.
-  const int second_transform_index = GetSubmapIndex(submaps[1]);
-  CHECK_LE(second_transform_index, next_transform_index);
+  const int second_submap_index = GetSubmapIndex(insertion_submaps[1]);
+  CHECK_LE(second_submap_index, next_submap_index);
   // Extrapolate if necessary.
-  if (second_transform_index == next_transform_index) {
-    const auto& first_submap_transform =
-        submap_transforms_[first_transform_index];
-    submap_transforms_.push_back(
-        first_submap_transform *
-        sparse_pose_graph::ComputeSubmapPose(*submaps[0]).inverse() *
-        sparse_pose_graph::ComputeSubmapPose(*submaps[1]));
+  if (second_submap_index == next_submap_index) {
+    const auto& first_submap_pose =
+        optimization_problem_.submap_data().at(first_submap_index).pose;
+    optimization_problem_.AddSubmap(
+        submap_states_[first_submap_index].trajectory,
```

Comment:
nit: Use `second_submap_index` for readability, even though both trajectories are the same. Similar in 3D.